### ISSUE TITLE
We don't have test coverage for aarch64 (ARM64) bfloat16 feature (__ARM_FEATURE_BF16)

### ISSUE THREAD

**Issue by swolchok** [pytorch/pytorch#142703]
### 🐛 Describe the bug

As demonstrated by #142370 and #142501, we have no builds that will catch compilation failures under `__ARM_FEATURE_BF16`, let alone running tests in this configuration.

### Versions

N/A

**Comment by swolchok** [comment]
Builds should be as simple as adding -march=armv8.2-a+bf16 to linux-aarch64 build flags (in a separate configuration). For tests, we'd need runners with hardware support, either M2 or later Macs or AWS Graviton 3 or later.

**Comment by swolchok** [comment]
> simple

I don't know how to set this up and it looks like .ci/pytorch/build.sh both has complicated control flow and doesn't have existing support for adding to build flags, but surely it can't be that complicated?

**Comment by swolchok** [comment]
> render binaries we produce useless on HW platforms that do not support that feature

yes, but that doesn't matter. we could happily throw the results away, it's just that we'd have some sort of automated guarantee we don't break the build on these platforms.

> add the compile flag

#142760 tries this, but didn't work. not sure why yet.

**Comment by swolchok** [comment]
Locally trying a build on M1 Mac with `env CFLAGS=-march=armv8.2-a+bf16 CXXFLAGS=-march=armv8.2-a+bf16 python setup.py develop`, and seeing ICE like
```
fatal error: error in backend: Cannot select: 0x149be1bb0: bf16 = uint_to_fp 0x149be1ec0
  0x149be1ec0: i32 = and 0x149be2240, Constant:i32<65535>
    0x149be2240: i32,ch = CopyFromReg 0x149796540, Register:i32 %33
      0x149be1d70: i32 = Register %33
    0x149be1c90: i32 = Constant<65535>
In function: _ZN2at6native12_GLOBAL__N_110add_kernelERNS_18TensorIteratorBaseERKN3c106ScalarE
c++: error: clang frontend command failed with exit code 70 (use -v to see invocation)
```
in `pytorch/build/aten/src/ATen/UfuncCPUKernel_add.cpp.DEFAULT.cpp`. also observed in vec_test_all_types.cpp. Not sure why (would be interested in whether it repros on an M2+ Mac); doesn't bode well for using Macs to do this testing (and also we should attempt to fix it).

**Comment by swolchok** [comment]
For M2, the appropriate -march would apparently be `-march=armv8.6a`, so perhaps I'm just unnecessarily hitting an unused case in the Apple clang backend.

**Comment by swolchok** [comment]
> For M2, the appropriate -march would apparently be `-march=armv8.6a`, so perhaps I'm just unnecessarily hitting an unused case in the Apple clang backend.

Failure persists with `-march=armv8.6a`.

**Comment by swolchok** [comment]
Found a bug in vec128_bfloat_neon.h that looks like it explains the ICE; confirming fix and will send patch.

**Comment by malfet** [comment]
I wonder if I've just enabled the coverage by https://github.com/pytorch/pytorch/pull/143129

**Comment by swolchok** [comment]
> I wonder if I've just enabled the coverage by #143129

not certain, but I think the issue fixed by #142879 should break vec_test_all_types in an aarch64+bf16 build

**Comment by ZainRizvi** [comment]
Hi @swolchok , is the challenge here actually adding the tests to the aarch64 CI? The machine types should be available, so please let us know what is blocking you from being able to add the appropriate tests.

**Comment by swolchok** [comment]
Status update on support here: I've discovered that older compiler versions have spotty support for bfloat16_t. 

- clang versions before 15 won't compile bfloat16 at all, and we use clang 14 to compile our Mac wheels currently.
- clang15 fails to build PyTorch on linux-aarch64 with -march=armv8.6-a due to instruction selection failures (examples at the bottom of this post); working on getting a clean repro together, but I fear that the solution is going to be to gate to a newer compiler version.

We need to do the following things:
- find the oldest compiler version that actually works on Linux and Mac for gcc and clang. (it looks like we don't do Windows aarch64 builds at all, so aarch64+bf16 can be punted there for now.) gcc 11.4.1 appears to be good enough to compile successfully, but I haven't tried older versions.
- decide (and/or remind me) what our support is going to look like -- do we *only* support the oldest compiler version, or do we also take advantage of features available if you aren't stuck on old OS releases? What compiler versions do we support on each platform? This matters because bfloat16 support is broken in different ways on older clang and gcc versions.
- update __ARM_FEATURE_BF16 gating accordingly.
- add a matrix of CI jobs to make sure __ARM_FEATURE_BF16 code is built and tested, presumably both on Linux and Mac, with the appropriate compiler versions.

Example clang15 ICE:
```
fatal error: error in backend: Cannot select: 0x39e2b1b8: v8bf16 = AArch64ISD::UZP1 0x2ed028d8, 0x30834230
  0x2ed028d8: v8bf16,ch = CopyFromReg 0x3b6f1788, Register:v8bf16 %30
    0x2ecfe3f8: v8bf16 = Register %30
  0x30834230: v8bf16,ch = CopyFromReg 0x3b6f1788, Register:v8bf16 %34
    0x30836990: v8bf16 = Register %34
In function: _ZZZZN2at6native12_GLOBAL__N_131grid_sampler_2d_cpu_kernel_implERKNS_10TensorBaseES4_S4_llbENK3$_0clEvENKUlvE2_clEvENKUlllE_clEll
```

Second clang15 ICE:
```
fatal error: error in backend: Cannot select: 0x35264748: bf16 = select 0x300ec368, ConstantFP:bf16<APFloat(65535)>, ConstantFP:bf16<APFloat(0)>
  0x300ec368: i32 = and 0x3f75c6f0, Constant:i32<1>
    0x3f75c6f0: i32 = extract_vector_elt 0x35263df0, Constant:i64<0>
      0x35263df0: v4i32 = insert_subvector undef:v4i32, 0x3cc2b360, Constant:i64<0>
        0x25ee7d20: v4i32 = undef
        0x3cc2b360: v2i32 = xor 0x3f933f00, 0x2cf1a778
          0x3f933f00: v2i32 = or 0x3f75cdd8, 0x3ad0c848
            0x3f75cdd8: v2i32 = AArch64ISD::FCMLTz 0x3e0981f8
              0x3e0981f8: v2f32 = bitcast 0x3f9347f0
                0x3f9347f0: v2i32 = AArch64ISD::VSHL 0x3cc2bec0, Constant:i32<16>
                  0x3cc2bec0: v2i32 = EXTRACT_SUBREG 0x2cf1a9e8, TargetConstant:i32<2>


                  0x3f933fd0: i32 = Constant<16>
            0x3ad0c848: v2i32 = AArch64ISD::FCMGEz 0x3e0981f8
              0x3e0981f8: v2f32 = bitcast 0x3f9347f0
                0x3f9347f0: v2i32 = AArch64ISD::VSHL 0x3cc2bec0, Constant:i32<16>
                  0x3cc2bec0: v2i32 = EXTRACT_SUBREG 0x2cf1a9e8, TargetConstant:i32<2>


                  0x3f933fd0: i32 = Constant<16>
          0x2cf1a778: v2i32 = BUILD_VECTOR Constant:i32<-1>, Constant:i32<-1>
            0x262f3b58: i32 = Constant<-1>
            0x262f3b58: i32 = Constant<-1>
        0x3e0985a0: i64 = Constant<0>
      0x3e0985a0: i64 = Constant<0>
    0x35263f90: i32 = Constant<1>
  0x28664c08: bf16 = ConstantFP<APFloat(65535)>
  0x3f75c7c0: bf16 = ConstantFP<APFloat(0)>
In function: _ZN2at6native7DEFAULT15vectorized_loopIRZZZNS1_L12angle_kernelERNS_18TensorIteratorBaseEENK4$_28clEvENKUlvE3_clEvEUlN3c108BFloat16EE_RZZZNS1_L12angle_kernelES4_ENKS5_clEvENKS6_clEvEUlNS_3vec7DEFAULT10VectorizedIS8_EEE_EEvPPcllOT_OT0_
```

**Comment by swolchok** [comment]
Just to cite an existence proof, I've found that clang 17, 18, and 19 can successfully build PyTorch for linux-aarch64 with CFLAGS="-march=armv8.6-a -Wno-missing-template-arg-list-after-template-kw" (and CXXFLAGS the same) if #143504 is applied. Trying 16 as well.

**Comment by swolchok** [comment]
clang16 has instruction selection failures when used to build for linux-aarch64 as above.

**Comment by malfet** [comment]
So, we have runners capable of executing the instruction, but compiler might be a challenge and moving the CD to a specific version of clang is probably a no-go, as clang does not produce manylinux2_28 compatible binaries, does it?
